ДЗ № 5.2 Сортировочная станция, Шестаков Николай

NicholayShestakov opened the pull request:
Add
- Realization of sorting station
- Compile instruction
- Incorrect infix forms processing

NicholayShestakov committed:
Add stack library header file

NicholayShestakov committed:
Add stack library realisation file

NicholayShestakov committed:
Update stack header file commentaries

NicholayShestakov committed:
Update stack realisation file

NicholayShestakov committed:
Update stack realisation file

NicholayShestakov committed:
Update stack header file

NicholayShestakov committed:
Update stack header file

NicholayShestakov committed:
Update stack realisation file

NicholayShestakov committed:
Update stack header file

NicholayShestakov committed:
Update stack realisation file

NicholayShestakov committed:
Add sorting station algorithm

NicholayShestakov committed:
Add compile instruction for sorting station

NicholayShestakov committed:
Update .gitignore with adding ignoring CMake build directory

NicholayShestakov committed:
Add CMakeLists.txt for build sorting station

NicholayShestakov committed:
Update instructions how to build

WoWaster commented on CMakeLists.txt:
Вот это явно придётся и в 5.1 затащить

WoWaster commented on src/stack_and_queue/stack/README.md:
```diff
@@ -0,0 +1,3 @@
+## Compile instruction for sorting station
+cmake . -B build\
+cmake --build build
```
~~~suggestion
```console
$ cmake . -B build
$ cmake --build build
```
~~~

Изучение Markdown может сильно упросить жизнь. Он правда простой!

WoWaster commented on src/stack_and_queue/stack/stack.c:
Здесь замечания такие же

WoWaster commented on src/stack_and_queue/stack/sorting_station.c:
```diff
@@ -0,0 +1,95 @@
+// Takes infix form of expression as string and returns pointer on postfix form string.
+char* sortingStation(char* infixForm)
+{
+    char* output = calloc(strlen(infixForm), sizeof(output) + 1);
```
На самом деле, по-моему, всё-таки надо выделять `2 * strlen(infixForm)`, чтобы оно точно влезло.

WoWaster commented on src/stack_and_queue/stack/sorting_station.c:
```diff
@@ -0,0 +1,95 @@
+    int indexOfOutput = 0;
+    for (int indexOfInfix = 0; infixForm[indexOfInfix] != '\0'; ++indexOfInfix) {
+        if (isdigit(infixForm[indexOfInfix])) {
+            output[indexOfOutput] = infixForm[indexOfInfix];
+            output[indexOfOutput + 1] = ' ';
+            indexOfOutput += 2;
```
```suggestion
            output[indexOfOutput++] = infixForm[indexOfInfix];
            output[indexOfOutput++] = ' ';
```
но это косметика, выбирайте, как нравится

WoWaster commented on src/stack_and_queue/stack/sorting_station.c:
```diff
@@ -0,0 +1,95 @@
+}
+
+// Takes infix form of expression as string and returns pointer on postfix form string.
+char* sortingStation(char* infixForm)
```
Есть ощущение, что Вы на какие-то ошибки (типа небалансности скобок) забиваете, это нужно поправить.

WoWaster commented on src/stack_and_queue/stack/sorting_station.c:
```diff
@@ -0,0 +1,95 @@
+        if (isOperatorFirstPriority(infixForm[indexOfInfix]) || infixForm[indexOfInfix] == '(') {
+            push(operators, infixForm[indexOfInfix]);
+        }
+        if (isOperatorSecondPriority(infixForm[indexOfInfix])) {
```
Что-то мне здесь не нравится. Контр-пример сходу не приведу, но точно понимаю, что Вам все-таки важно как-то сравнивать приоритет операторов.

WoWaster reviewed (changes requested):
Тут тоже всё довольно круто!

NicholayShestakov committed:
Update grammary in documentations

NicholayShestakov committed:
Update stack realization

Remove pop and peek -1 return if stack is empty.
Add asserts for memory allocation.
Add asserts for not empty stack in pop and peek.

NicholayShestakov committed:
Update stack header documentation

NicholayShestakov committed:
Add build instruction in README

NicholayShestakov committed:
Delete build instruction from stack folder

NicholayShestakov committed:
Update sorting station

Add error messages for incorrect infix forms.

WoWaster reviewed (approved):
Мне, честно, уже просто лень придумывать контр-пример, так что зачтено.
